Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+on the CPU cores in a very specific order.
+
+A classical race condition example is an incorrect implementation for
+a release operation of a resource counter:
+
+.. slide:: Race condition: resource counter release
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+      void release_resource()
+      {
+           counter--;
+
+	   if (!counter)
+               free_resource();
+      }
+
+
+A resource counter is used to keep a shared resource available until
+the last user releases it but the above implementation has a race
+condition that can cause freeing the resource twice:
+
+
+.. slide:: Race condition scenario
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+                             counter is 2
+
+        Thread A                                Thread B
+
+           *
+           |
+           |
+     +---------------------+
+     |  dec counter        |   counter is 1
+     |  cEEE        	   |
+     +---------------------+
+           |
+           |                 B preempts A
+           +-----------------------------------------------+
+                                                           |
+	                                                   v
+			   			    +----------------------+
+                             counter is 0           | dec counter  	   |
+                                                    | if (!counter)	   |
+                           resource is freed        |     free_resource(); |
+			   			    | cEEE    	       	   |
+			   			    +----------------------+
+                        B finishes, A continues            |
+           +-----------------------------------------------+
+           |
+           v
+    +----------------------+
+    | if (!counter)	   |
+    |     free_resource(); | resource is freed
+    | cEEE     	       	   |
+    +----------------------+
+
+In most cases the `release_resource()` function will only free the
+resource once. However, in the scenario above, if thread A is
+preempted right after decrementing `counter` and thread B calls
+`release_resource()` it will cause the resource to be freed. When
+resumed, thread A will also free the resource since the counter value
+is 0.
+
+To avoid race conditions the programmer must first identify the
+critical section that can generate a race condition. The critical
+section is the part of the code that reads and writes shared memory
+from multiple parallel contexts.
+
+In the example above, the minimal critical section is starting with
+the counter decrement and ending with checking the counter's value.
+
+Once the critical section has been identified race conditions can be
+avoided:
+
+.. slide:: Avoiding race conditions
+   :inline-contents: True
+   :level: 2
+
+   * make the critical section **atomic** (so that preemption can not
+     occur in the critical section)
+
+   * **disable preemption** while the critical section is executing
```

Comment:
i.e. the operating system will not preempt the current execution context and replace it with another execution context